packages/world/src/systems/MineSystem.sol

```diff
@@ -217,6 +221,52 @@ contract MineSystem is System {
       }
     }
   }
+
+  // TODO: this is ugly, but doing this to avoid stack too deep errors. We should refactor later.
+  function _processMassReduction(EntityId caller, Vec3 callerCoord, uint16 toolSlot, EntityId mined)
+    internal
+    returns (uint128, bool)
+  {
+    ToolData memory toolData = InventoryUtils.getToolData(caller, toolSlot);
+    (uint128 finalMass, uint128 toolMassReduction, uint128 energyReduction) = _getMassReduction(toolData, mined);
+
+    if (energyReduction > 0) {
+      // If player died, return early
+      (uint128 callerEnergy,) = transferEnergyToPool(caller, energyReduction);
+      if (callerEnergy == 0) {
+        return (finalMass, false);
+      }
+    }
+
+    // Apply tool usage after decreasing player energy so we make sure the player is alive
+    toolData.applyMassReduction(callerCoord, toolMassReduction);
+    return (finalMass, true);
+  }
+
+  function _getMassReduction(ToolData memory toolData, EntityId mined)
+    internal
+    view
+    returns (uint128, uint128, uint128)
+  {
+    uint128 massLeft = Mass._getMass(mined);
+    if (massLeft == 0) {
+      return (0, 0, 0);
+    }
+
+    uint128 toolMassReduction = toolData.getMassReduction(massLeft);
+
+    // if tool mass reduction is not enough, consume energy from player up to mine energy cost
+    uint128 energyReduction = 0;
+    if (toolMassReduction < massLeft) {
+      uint128 remaining = massLeft - toolMassReduction;
+      energyReduction = MINE_ENERGY_COST <= remaining ? MINE_ENERGY_COST : remaining;
+      massLeft -= energyReduction;
+    }
+
+    uint128 finalMass = massLeft - toolMassReduction;
```

Comment:
can this underflow?

Reply:
no, because we pass massLeft to `getMassReduction`, which limits the amount of mass that can be reduced

Reply:
ahh and then also when we subtract from massLeft in the player case we're using remaining, nice

Reply:
also in lines 259-264, we only reduce massLeft up to `massLeft - toolMassReduction`, so it can never be smaller than `toolMassReduction`